projects/plugins/jetpack/extensions/blocks/sharing-button/sharing-sources-block.php

```diff
@@ -0,0 +1,1397 @@
+
+		parent::redirect_request( $pocket_url );
+	}
+}
+
+/**
+ * Telegram sharing service.
+ */
+class Share_Telegram_Block extends Sharing_Source_Block {
+	/**
+	 * Service short name.
+	 *
+	 * @var string
+	 */
+	public $shortname = 'telegram';
+
+	/**
+	 * Service name.
+	 *
+	 * @return string
+	 */
+	public function get_name() {
+		return __( 'Telegram', 'jetpack' );
+	}
+
+	/**
+	 * Process sharing request. Add actions that need to happen when sharing here.
+	 *
+	 * @param WP_Post $post Post object.
+	 * @param array   $post_data Array of information about the post we're sharing.
+	 *
+	 * @return void
+	 */
+	public function process_request( $post, array $post_data ) {
+		// Record stats
+		parent::process_request( $post, $post_data );
+
+		$telegram_url = esc_url_raw( 'https://telegram.me/share/url?url=' . rawurlencode( $this->get_share_url( $post->ID ) ) . '&text=' . rawurlencode( $this->get_share_title( $post->ID ) ) );
+
+		parent::redirect_request( $telegram_url );
+	}
+}
+
+/**
+ * WhatsApp sharing service.
+ */
+class Jetpack_Share_WhatsApp_Block extends Sharing_Source_Block {
+	/**
+	 * Service short name.
+	 *
+	 * @var string
+	 */
+	public $shortname = 'jetpack-whatsapp';
+
+	/**
+	 * Service name.
+	 *
+	 * @return string
+	 */
+	public function get_name() {
+		return __( 'WhatsApp', 'jetpack' );
+	}
+
+	/**
+	 * Process sharing request. Add actions that need to happen when sharing here.
+	 *
+	 * @param WP_Post $post Post object.
+	 * @param array   $post_data Array of information about the post we're sharing.
+	 *
+	 * @return void
+	 */
+	public function process_request( $post, array $post_data ) {
+		// Record stats
+		parent::process_request( $post, $post_data );
+
+		// Firefox for desktop doesn't handle the "api.whatsapp.com" URL properly, so use "web.whatsapp.com"
+		if ( User_Agent_Info::is_firefox_desktop() ) {
+			$url = 'https://web.whatsapp.com/send?text=';
+		} else {
+			$url = 'https://api.whatsapp.com/send?text=';
+		}
+
+		$url .= rawurlencode( $this->get_share_title( $post->ID ) . ' ' . $this->get_share_url( $post->ID ) );
+
+		parent::redirect_request( $url );
+	}
+}
+
+/**
+ * Mastodon sharing service.
+ */
+class Share_Mastodon_Block extends Sharing_Source_Block {
+	/**
+	 * Service short name.
+	 *
+	 * @var string
+	 */
+	public $shortname = 'mastodon';
+
+	/**
+	 * Service name.
+	 *
+	 * @return string
+	 */
+	public function get_name() {
+		return __( 'Mastodon', 'jetpack' );
+	}
+
+	/**
+	 * Process sharing request. Add actions that need to happen when sharing here.
+	 *
+	 * @param WP_Post $post Post object.
+	 * @param array   $post_data Array of information about the post we're sharing.
+	 *
+	 * @return void
+	 */
+	public function process_request( $post, array $post_data ) {
+		if ( empty( $_POST['jetpack-mastodon-instance'] ) ) {
+			require_once WP_SHARING_PLUGIN_DIR . 'services/class-jetpack-mastodon-modal.php';
```

Comment:
This is something we'll have to be intentional about, to avoid errors if one day that file is removed or renamed.

What do you think about adding a comment in the file, mentioning that the `Jetpack_Mastodon_Modal` is used in the module as well as the block?

Reply:
What if I also move this file into `block` directory? To keep things fully separated? What do you think?